### DISCUSSION TITLE
How to consume messages, of the same type, in different topics?

### DISCUSSION THREAD

**Discussion by gabriel-sk-dev** [BEagle1984/silverback#182] [Q&A] [answered]
Lets say we hay the following topics:
- demo1
- demo2

And we are publishing the same message in the topics. (for example: MySampleMessage)

How to consume the messages in these topics (2 subscribers with same message type) within the same application?

**Comment by BEagle1984** [comment, marked as answer]
You have basically 3 options (possibly more):
1. Create two message models that inherit from `MySampleMessage`, for example `MySampleMessageFromDemo1` and `MySampleMessageFromDemo2`. Then configure the inbound endpoints to deserialize those types instead of the base `MySampleMessage`. The JSON is obviously compatible and you now have two different messages to subscribe to.
2. Subscribe once to `IInboundEnvelope<MySampleMessage>` and in the subscriber you can check the source topic (`envelope.Endpoint.Name`) and call two different methods (directly or via the `IPublisher`, wrapping the message inside another message model).
3. Build an `IMessageFilterAttribute` to decorate your subscribers and filter the messages from a specific topic. You can use the [KafkaGroupIdFilterAttribute](https://github.com/BEagle1984/silverback/blob/master/src/Silverback.Integration.Kafka/Messaging/Subscribers/KafkaGroupIdFilterAttribute.cs) as inspiration. (see also https://silverback-messaging.net/concepts/broker/kafka/multiple-consumer-groups.html).

I'd prefer the first choice, to avoid polluting my code and keep all transport-related configurations in a central place (the endpoint configurator).
To be fair, the subscriber filters can be specified in the `AddSubscriber` call instead of being added on the subscriber method directly  (see the note and example here: https://silverback-messaging.net/concepts/broker/kafka/multiple-consumer-groups.html) but on the other hand, it wouldn't be immediately visible in the subscriber that you are indeed filtering the incoming messages and for this reason, I'd still prefer the first choice in most cases.

**Reply by gabriel-sk-dev** [reply]
Thanks @BEagle1984

I used the decorator filter strategy and work's fine.
